feat(#29): work-in-progress plugin

bartoszmajsak opened the pull request:
Introduces status update based on PR title prefix.

Fixes #25

bartoszmajsak commented on plugin/internal/test/test_doubles.go:
```diff
@@ -0,0 +1,46 @@
+	"encoding/json"
+)
+// nolint
+func EventPayload(payloadFile string) []byte {
```
This function is used in tests and was there before - I just extracted it to be re-used in other tests in this PR. Do you think naming it `FileToBytes` would make it easier to read tests?

MatousJobanek replied:
I know that it had been part of the `event_hadler_test.go` in `test-keeper` plugin before. I remember when I saw it for the first time, I was pretty confused by the method name but as it was part of the particular test file, then I wasn't complaining about it. Now, when it is moved to a common file and used as a "public util function" I would welcome some name that says what it really does.

bartoszmajsak replied:
I understand your point of view, however, there are two things which keep me defending this choice:

- it's an internal package intended to be used only in tests (that's also the reason why you see `// nolint` here. Could be that I should add documentation here.
- I prefer to read : 

```golang
handler.HandleEvent(githubevents.PullRequest, guid, EventPayload("test_fixtures/github_calls/wip_pr_opened.json"))
``` 
than 
```golang
handler.HandleEvent(githubevents.PullRequest, guid, FileToBytes("test_fixtures/github_calls/wip_pr_opened.json"))
```

somehow it says exactly what it does.

One can argue to have:

```golang
eventPayload := FileToBytes("test_fixtures/github_calls/wip_pr_opened.json")
handler.HandleEvent(githubevents.PullRequest, guid, eventPayload)
```

but I'm not sure we win anything.

bartoszmajsak replied:
Went with the var split. Thanks for the discussion. Shipped with 9ccda31

bartoszmajsak commented on plugin/test-keeper/plugin/event_handler_test.go:
```diff
@@ -3,30 +3,18 @@ package plugin_test
-		nullLogger.Out = ioutil.Discard // TODO rethink if we want to discard logging entirely
-		logger = logrus.NewEntry(nullLogger)
-	})
+var _ = Describe("WIP Plugin features", func() {
```
Right, renamed it in the wrong place ;)

bartoszmajsak replied:
Fixed in 9b7e557

bartoszmajsak commented on plugin/work-in-progress/main.go:
```diff
@@ -0,0 +1,40 @@
+
+func helpProvider(enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	return &pluginhelp.PluginHelp{
+		Description: `Work-in-progress plugin`,
```
Why?

MatousJobanek replied:
I'm stupid - I read it `Work-in-progress-plugin` missed the space in front of the "plugin" word. So I thought that you used a new format using the dash instead of space :-). Sorry ;-)

bartoszmajsak replied:
Well it's just a text :)

bartoszmajsak commented on plugin/work-in-progress/plugin/event_handler_unit_test.go:
```diff
@@ -0,0 +1,61 @@
+				Expect(handler.IsWorkInProgress(&title)).To(BeFalse())
+			},
+			Entry("regular PR title", "fix(#1): off-by one bug"),
+			Entry("wip", "wip-fix off-by one bug"),
```
I'm open to any suggestion.

bartoszmajsak replied:
Does "not a regular wip prefix" looks better? Added in 278ac9a

MatousJobanek replied:
"wip without subsequent space"

MatousJobanek replied:
> Does "not a regular wip prefix" looks better?

yeah, that is fine

bartoszmajsak replied:
> "wip without subsequent space"

This you can reason from the string afterwards?

```golang
Entry("not a supported wip prefix", "wip-fix off-by one bug"),
```

bartoszmajsak commented on plugin/work-in-progress/plugin/event_handler_unit_test.go:
```diff
@@ -0,0 +1,61 @@
+
+		It("should not recognize PR as work-in-progress if title doesn't start with WIP", func() {
+
+		})
```
![headshot](https://user-images.githubusercontent.com/719616/37090369-6514aee0-2204-11e8-9ded-010d76f79fe4.gif)

Right, fixed in ab26d43

bartoszmajsak commented on plugin/work-in-progress/plugin/even_handler_suite_test.go:
```diff
@@ -0,0 +1,12 @@
+	. "github.com/onsi/gomega"
+)
+
+func TestTestKeeper(t *testing.T) {
```
Right, but you could as well just edit it ;) Fixed in 9926189

bartoszmajsak commented on plugin/work-in-progress/plugin/event_handler_gh_test.go:
```diff
@@ -0,0 +1,106 @@
+				Reply(201) // This way we implicitly verify that call happened after `HandleEvent` call
+
+			// when
+			handler.HandleEvent(githubevents.PullRequest, "random", EventPayload("test_fixtures/github_calls/ready_pr_opened.json"))
```
Ah, very good catch :) It should be applied to all the tests. I will take care of it.

How about 

```golang
// when
			err := handler.HandleEvent(githubevents.PullRequest, "random", EventPayload("test_fixtures/github_calls/ready_pr_opened.json"))

			// then - implicit verification of /statuses call occurrence with proper payload
			Expect(err).To(BeNil())
```

MatousJobanek replied:
exactly, as I did in my PR https://github.com/arquillian/ike-prow-plugins/pull/20/files#diff-b1b7ef7199ea998477e0667d3d9edcdfR53

bartoszmajsak replied:
With a small difference of `Should` vs `To`, but it's a detail. Fixed in 3b7788f

MatousJobanek commented on plugin/work-in-progress/plugin/event_handler.go:
```diff
@@ -0,0 +1,79 @@
+			Description: &reason,
+		}); err != nil {
+			gh.Log.Info("Error handling event.", err)
+		}
```
For this, you can use the `CommitScmService` so it will look like:
```java
head := scm.CreateScmCommitService(gh.Client, gh.Log, event.Repo, *event.PullRequest.Head.SHA)
if gh.IsWorkInProgress(event.PullRequest.Title) {
	head.Fail("PR is in progress and can't be merged yet. You might want to wait with review as well")
}
head.Success("PR is ready for review and merge")
```

bartoszmajsak replied:
My understanding is that we are going to refactor this piece together tomorrow, so shall I do it here or at the time of refactoring?

MatousJobanek replied:
If we don't forget tomorrow, then we can keep it as it is :-).
I just thought that it would be good to have it unified so it is easier to refactor everything together

bartoszmajsak replied:
Fair point. Doing it right now

bartoszmajsak commented:
> And another thing to ask - previously, the nullLogger was created in BeforeSuite, so I moved it to the suite definition https://github.com/arquillian/ike-prow-plugins/blob/master/plugin/test-keeper/plugin/test_keeper_suite_test.go#L16-L22 this var is then used in the tests.
In your PR, you create the logger in BeforeEach. which version is better? WDYT?

I unified it in one place, but forgot about the other. It's now fixed in b81e0fc. Thanks for pointing it out.
